pkg/controller/resourcedistribution/resourcedistribution_controller.go

```diff
@@ -0,0 +1,391 @@
+			newObj := e.ObjectNew.(*appsv1alpha1.ResourceDistribution)
+			if !reflect.DeepEqual(oldObj.Spec, newObj.Spec) {
+				klog.V(3).Infof("Observed updated Spec for ResourceDistribution: %s/%s", oldObj.Namespace, newObj.Name)
+				return true
+			}
+			return false
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to all namespaces
+	if err = c.Watch(&source.Kind{Type: &corev1.Namespace{}}, &enqueueRequestForNamespace{reader: mgr.GetCache()}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileResourceDistribution{}
+
+// ReconcileResourceDistribution reconciles a ResourceDistribution object
+type ReconcileResourceDistribution struct {
+	client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+//+kubebuilder:rbac:groups=apps.kruise.io,resources=resourcedistributions,verbs=get;list;watch;
+//+kubebuilder:rbac:groups=apps.kruise.io,resources=resourcedistributions/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch;
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;delete
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ResourceDistribution object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *ReconcileResourceDistribution) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	klog.V(3).Infof("ResourceDistribution(%s) begin to reconcile\n\n", req.NamespacedName.Name)
+
+	// fetch resourcedistribution instance as distributor
+	distributor := &appsv1alpha1.ResourceDistribution{}
+	if err := r.Client.Get(context.TODO(), req.NamespacedName, distributor); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// do reconcile
+	newStatus, err := r.doReconcile(distributor)
+
+	// update distributor status
+	if newStatus == nil || reflect.DeepEqual(distributor.Status, newStatus) {
+		return ctrl.Result{}, err
+	}
+	distributor.Status = *newStatus
+	if err := r.updateDistributorStatus(distributor); err != nil {
+		klog.Errorf("ResourceDistribution update status error, err: %v, name: %s, new status %v", err, distributor.Name, newStatus)
+		return reconcile.Result{}, err
+	}
+
+	return ctrl.Result{}, err
```

Comment:
`return ctrl.Result{}, nil`

Reply:
This err indicates the error during resource distribution.
If no error occurred during resource distribution, this err must be nil, but we cannot return nil directly.
Maybe we should make this logic more clear:
```go
        // update distributor status
	if newStatus == nil || reflect.DeepEqual(distributor.Status, newStatus) {
		return ctrl.Result{}, err
	}
	distributor.Status = *newStatus
	if updateStatusErr := r.updateDistributorStatus(distributor); updateStatusErr != nil {
		klog.Errorf("ResourceDistribution update status error, err: %v, name: %s, new status %v", updateStatusErr, distributor.Name, newStatus)
		return reconcile.Result{}, fmt.Errorf("distribution error: %v, update status error: %v", err, updateStatusErr)
	}

	// return distribution err, if no error occurred, this err must be nil
	return ctrl.Result{}, err
```

Reply:
According to the logic in `doReconcile`, the `err` here can only be nil if `newStatus` is not nil?

Reply:
No, because we make a error in line 290, the `err` here may not be nil.
```
if numberOfErrs != 0 {
    err = fmt.Errorf("some unexpected errors occurred when distributing resource, error messages has been recorded in .status.conditions, name: %s", distributor.Name)
}
```